Improve docs about using Logger (#1441)
nestjs/docs.nestjs.com: content/techniques/logger.md

```typescript
import { Logger, Injectable } from '@nestjs/common';

@Injectable()
class MyService {
  private readonly logger = new Logger(MyService.name);
  
  doSomething() {
    this.logger.log('Doing something...');
  }
}
```

In the default logger implementation, `context` is printed in the square brackets, like `NestFactory` in the example below:

```bash
[Nest] 19096   - 12/08/2019, 7:12:59 AM   [NestFactory] Starting Nest application...
```

If we supply a custom logger via `app.useLogger()`, it will actually be used by Nest internally. That means that our code remains implementation agnostic, while we can easily substitute the default logger for our custom one by calling `app.useLogger()`.

That way if we follow the steps from the previous section and call `app.useLogger(app.get(MyLogger))`, the following calls to `this.logger.log()` from `MyService` would result in calls to method `log` from `MyLogger` instance.

This should be suitable for most cases. But if you need more customization (like adding and calling custom methods), move to the next section. 

#### Injecting a custom logger

> 

Reviewer: ```suggestion
```

```typescript
import { Logger, Injectable } from '@nestjs/common';

@Injectable()
class MyService {
  private readonly logger = new Logger(MyService.name);
  
  doSomething() {
    this.logger.log('Doing something...');
  }
}
```

In the default logger implementation, `context` is printed in the square brackets, like `NestFactory` in the example below:

```bash
[Nest] 19096   - 12/08/2019, 7:12:59 AM   [NestFactory] Starting Nest application...
```

If we supply a custom logger via `app.useLogger()`, it will actually be used by Nest internally. That means that our code remains implementation agnostic, while we can easily substitute the default logger for our custom one by calling `app.useLogger()`.

That way if we follow the steps from the previous section and call `app.useLogger(app.get(MyLogger))`, the following calls to `this.logger.log()` from `MyService` would result in calls to method `log` from `MyLogger` instance.

This should be suitable for most cases. But if you need more customization (like adding and calling custom methods), move to the next section. 

#### Injecting a custom logger

> Sometimes you may want to add custom methods to your logger class.
> In order to use these methods you'll have to inject a custom logger to your services.
> We can use a [transient](/fundamentals/injection-scopes) scope for the logger so that each one of our services has its own custom context.
> If we maintain the `LoggerService` interface in our custom logger, we can still set our logger globally with `app.useLogger`.

Reviewer: ```suggestion
```

Alternatively you can print first initialization messages using default logger, and then switch to your custom one:

```typescript
const app = await NestFactory.create(ApplicationModule);
app.useLogger(app.get(MyLogger));
await app.listen(3000);
``` 

You can also inject this `MyLogger` provider in your feature classes, thus ensuring consistent logging behavior across both Nest system logging and application logging. See <a href="techniques/logger#using-the-logger-for-application-logging">Using the logger for application logging</a> and <a href="techniques/logger#injecting-a-custom-logger">Injecting a custom logger</a> below for more information.

#### Using the logger for application logging

We can combine several of the techniques above to provide consistent behavior and formatting across both Nest system logging and our own application event/message logging.

A good practice is to instantiate `Logger` class from `@nestjs/common` in each of our services. We can supply our service name as the `context` argument in the `Logger` constructor, like so:

```typescript
import { Logger, Injectable } from '@nestjs/common';

@Injectable()
class MyService {

>   logger = new Logger(MyService.name);

Reviewer: ```suggestion
  private readonly logger = new Logger(MyService.name);
```

>   logger = new Logger(MyService.name);

Reviewer 2: It would be better to show this in the `constructor` to better follow dependency injection principles.

Author: Do you mean like this?
```
constructor() {
  this.logger = new Logger(MyService.name);
}
```

From my standpoint it is no different from what wrote, but through the times of my googling I only saw variant with property initializer (even in the nestjs source), so I used it here as well.

Reviewer 2: Rather something like
```ts
class MyService {
  constructor(private readonly logger: Logger) {
    this.logger.setContext(MyService.name);
  }
}
```

Author: I had some issues with that with `nestjs-pino` because it has its own provider for `Logger`, so this code actually injects `Logger` from `nestjs-pino` instead of `Logger` from `@nestjs/common`. And the former does not have `setContext` function. So it produces a runtime error. Maybe this is a mistake with the library, but still `new Logger` seems more safe.

Also, `setContext` mutates `Logger` instance and `Logger` seems to have singleton scope (https://github.com/nestjs/nest/blob/master/packages/common/services/logger.service.ts#L19), so I am not sure this would work.

Reviewer 2: Ah, in the case of singleton scopes, you'd be right about that. In the case of `nestjs-pino` I feel that's on the library. You could always get around that with `@Inject()`, but that's not what this PR is about ;). All right, LGTM.

  
  doSomething() {
    this.logger.log('Doing something...');
  }
}
```

In the default logger `context` is printed in the square brackets, like `NestFactory` in the example below:

> ```bash

Reviewer 2: Add a newline to separate the code block from the regular text.

Author: Done, thanks for noticing

> In the default logger `context` is printed in the square brackets, like `NestFactory` in the example below:

Reviewer: ```suggestion
In the default logger implementation, `context` is printed in the square brackets, like `NestFactory` in the example below:
```

[Nest] 19096   - 12/08/2019, 7:12:59 AM   [NestFactory] Starting Nest application...
```

> If we supply a custom logger via `app.useLogger`, it will actually be used by Nest automatically.
> That means that our code remains implementation agnostic, while we can easily substitute the default logger for our custom one by calling `app.useLogger`.
> That way if we follow the steps from previous section and call `app.useLogger(app.get(MyLogger))`, the following calls to `this.logger.log` from `MyService` would result in calls to method `log` from `MyLogger` instance.
> 
> This should be suitable for most cases. But if you need more customization for the Logger (like adding and calling custom methods), please follow to the next section. 

Reviewer: ```suggestion
If we supply a custom logger via `app.useLogger()`, it will actually be used by Nest internally. That means that our code remains implementation agnostic, while we can easily substitute the default logger for our custom one by calling `app.useLogger()`.

That way if we follow the steps from the previous section and call `app.useLogger(app.get(MyLogger))`, the following calls to `this.logger.log()` from `MyService` would result in calls to method `log` from `MyLogger` instance.

This should be suitable for most cases. But if you need more customization (like adding and calling custom methods), move to the next section. 
```

#### Injecting a custom logger

Sometimes you may want to add custom methods to your logger class.
In order to use these methods you'll have to inject a custom logger to your services.
We can use a [transient](/fundamentals/injection-scopes) scope for the logger so that each one of our services has its own custom context.
If we maintain the `LoggerService` interface in our custom logger, we can still set our logger globally with `app.useLogger`.

To start, extend the built-in logger with code like the following. We supply the `scope` option as configuration metadata for the `Logger` class, specifying a [transient](/fundamentals/injection-scopes) scope, to ensure that we'll have a unique instance of the `MyLogger` in each feature module. In this example, we do not extend the individual `Logger` methods (like `log()`, `warn()`, etc.), though you may choose to do so.

```typescript
import { Injectable, Scope, Logger } from '@nestjs/common';

@Injectable({ scope: Scope.TRANSIENT })
export class MyLogger extends Logger {

>   // Implement custom methods or extend default methods here,
>   // for example:

Reviewer: ```suggestion
```

The only downside of this solution is that your first initialization messages won't be printed anywhere at all, so in rare cases you may miss some important initialization errors.
Alternatively you can print first initialization messages using default logger, and then switch to your custom one:

```typescript
const app = await NestFactory.create(ApplicationModule);
app.useLogger(app.get(MyLogger));
await app.listen(3000);
``` 

You can also inject this `MyLogger` provider in your feature classes, thus ensuring consistent logging behavior across both Nest system logging and application logging. See <a href="techniques/logger#using-the-logger-for-application-logging">Using the logger for application logging</a> and <a href="techniques/logger#injecting-a-custom-logger">Injecting a custom logger</a> below for more information.

#### Using the logger for application logging

We can combine several of the techniques above to provide consistent behavior and formatting across both Nest system logging and our own application event/message logging.

> The best practice is to instantiate `Logger` class from `@nestjs/common` in each of our services. 
> We can supply our service name as the `context` argument in the `Logger` constructor, like so:

Reviewer: ```suggestion
A good practice is to instantiate `Logger` class from `@nestjs/common` in each of our services. We can supply our service name as the `context` argument in the `Logger` constructor, like so:
```

await app.listen(3000);
```

> Be mindful that if you supply `logger: false` to `NestFactory.create`, nothing will be logged until you call `useLogger`, so you may miss some important initialization errors.
> If you don't mind that some of your initial messages will be logged with the default logger, you can just omit the `logger: false` option.

Reviewer: ```suggestion
Be mindful that if you supply `logger: false` to `NestFactory.create`, nothing will be logged until you call `useLogger`, so you may miss some important initialization errors. If you don't mind that some of your initial messages will be logged with the default logger, you can just omit the `logger: false` option.
```
